### ISSUE TITLE
MAUI Picker always open up in Android 8 and below

### ISSUE THREAD

**Issue by sujith2ss** [dotnet/maui#15394]
### Description

Navigate from one page to another MAUI Picker always open, with out even selecting it on Android version 8 and below. Easily reproducible in Android Emulator.

### Steps to Reproduce

1. Clone the  repo https://github.com/sujith2ss/test
2. Build and run it on Android 8 device or emulator
3. Click on the Next Button on App
4. See the picker is opened already.

### Link to public reproduction project repository

https://github.com/sujith2ss/test

### Version with bug

7.0.49

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android 8 and below

**Comment by sujith2ss** [comment]
Demo video attached 
https://github.com/sujith2ss/test/blob/main/demo.mp4

**Comment by gktval** [comment]
I am experiencing this with Android versions 32/33. Have you found a workaround?

**Comment by sujith2ss** [comment]
@gktval I didn't notice this issue in andorid version higher than 8.
In android 8.0 or below if we try focus or unfocus the result are same.
In the example code you can see only page navigation to another page, but the picker is getting opened automatically.

**Comment by AndieH68** [comment]
Also seeing this behaviour in Android 13.0 with API 33

**Comment by AndieH68** [comment]
Also experiencing this on Android 13 API Version 33.

This occurs if the Picker is the first visible control on a page and the page regains focus from another page or a popup disappearing.

**Comment by thomasgalliker** [comment]
Yeah, there is some code in the PickerHandler.android.cs that says 'if focused, run click method'. I have no idea why this code is there. For most of the picker related handlers, I had to rewrite them from scratch because they were either not working or having memory leaks.

**Comment by Andy-Donegan** [comment]
Just to add same here as @AndieH68 Android 13 API version 33.

**Comment by XamlTest** [comment]
Verified this on Visual Studio Enterprise 17.9.0 Preview 1(8.0.3). Repro on Android 8.0-API26, not repro on Android 13.0-API33, Windows 11 with below Project:
[Sample.zip](https://github.com/dotnet/maui/files/13435992/Sample.zip)

**Comment by ViktorArbuzov** [comment]
any updates?

**Comment by donatellijl12** [comment]
I have this issue as well. Whenever navigating to a page with a picker on Android, the picker automatically opens. Is there a workaround other than building a custom control?

**Comment by Andy-Donegan** [comment]
> automatically

Yes there is a work around, I am away for the weekend but will post the work around I used and try to find the link for the fix so others can see it within here. Sorry but can not update until Monday.

**Comment by donatellijl12** [comment]
Placing the picker in a scroll view fixed it for me for on initial page navigation, but not whenever I navigate back to the page. I'm interested to see your work around when you are able to share. Thanks!

P.S. I also tried calling unfocus() in OnAppearing and listing the items in the xaml instead of using ItemsSource, neither which fixed the issue.

**Comment by Andy-Donegan** [comment]
I could not find the reference for how I came to this fix, as I did not create this somebody else figured this out and I have just used it.
In your .xaml page add this as the top element so it is target by the bug as the first element.
```
<!-- element added her for .Net 8 fix to stop popups auto opening on refocus to page.-->                
<Entry 
    x:Name="FakeEntry"
    Margin="0,0,0,0"
    HeightRequest="1"
    MaximumHeightRequest="1"
    TextColor="Black"
    Placeholder=""/>
```
in your code behind. You need to add the following because the Android Underline is present on all entry elements so the invisible element will still have an underline, below will remove it.
```
    // Complete below function is required to fix .Net 8 upgrade which breaks the picker class
    // Had to add invisible text input element also and then hide it (not invisible) to stop page focus causing
    // pickers to be opened all the time.
    protected override void OnHandlerChanged()
    {
        base.OnHandlerChanged();
#if ANDROID
        Android.Graphics.Drawables.GradientDrawable gd = new Android.Graphics.Drawables.GradientDrawable();
        gd.SetStroke(0, Android.Graphics.Color.Transparent);
        gd.SetColor(Android.Graphics.Color.Transparent);

        // Invisible Text Entry box for .Net 8 fix for popups auto being called on return to page focus.
        var view = FakeEntry.Handler.PlatformView as AndroidX.AppCompat.Widget.AppCompatEditText;
        view.Background = null;
        view.SetBackgroundColor(Android.Graphics.Color.Transparent);
        view.SetPadding(0, 0, 0, 0);
        view.SetCursorVisible(false);
        view.SetBackground(gd);
#endif
    }

All credit goes to somebody else.

**Comment by mchiz** [comment]
I created a more generic fix. I have been testing it and it seems to work.

First, you have to create this handler:
`public class PickerHandlerFixAndroidFocus : PickerHandler {
    public PickerHandlerFixAndroidFocus( ) {
        _onFocusChangeMethod = typeof( PickerHandler ).GetMethod( "OnFocusChange", BindingFlags.Instance | BindingFlags.NonPublic );
        _onClickMethod = typeof( PickerHandler ).GetMethod( "OnClick", BindingFlags.Instance | BindingFlags.NonPublic );

        System.Diagnostics.Debug.Assert( _onFocusChangeMethod != null && _onClickMethod != null );
    }

    protected override void ConnectHandler( MauiPicker platformView ) {
        base.ConnectHandler( platformView );

        var focusChangeDelegate = ( System.EventHandler< Android.Views.View.FocusChangeEventArgs > )Delegate.CreateDelegate( typeof( System.EventHandler<Android.Views.View.FocusChangeEventArgs> ), this, _onFocusChangeMethod );
        var clickDelegate = ( System.EventHandler )Delegate.CreateDelegate( typeof( System.EventHandler ), this, _onClickMethod );

        platformView.Click -= clickDelegate;
        platformView.FocusChange -= focusChangeDelegate;
        
        platformView.Click += OnClick;
        platformView.FocusChange += OnFocusChange;
    }

	void OnClick( object? sender, EventArgs e ) {
        if( _ignoreNextClick ) {
            _ignoreNextClick = false;
            return;
        }

        _onClickMethod!.Invoke( this, [ sender, e ] );
    }

	void OnFocusChange( object? sender, global::Android.Views.View.FocusChangeEventArgs e ) {
        if( e.HasFocus ) {
            _ignoreNextClick = true;
            return;
        }

        _ignoreNextClick = false;

        _onFocusChangeMethod!.Invoke( this, [ sender, e ] );
	}

    bool _ignoreNextClick = false;
    MethodInfo? _onClickMethod;
    MethodInfo? _onFocusChangeMethod;
}
`

And then you just register it in your MauiProgram.cs (Android only):

`var builder = MauiApp.CreateBuilder( );

builder.ConfigureMauiHandlers( cf => {
                #if ANDROID
                cf.AddHandler( typeof( Picker ), typeof( PickerHandlerFixAndroidFocus ) );
                #endif
            } )`

**Comment by RichardMarsh** [comment]
> Also experiencing this on Android 13 API Version 33.
> 
> This occurs if the Picker is the first visible control on a page and the page regains focus from another page or a popup disappearing.

I had the same problem when closing a popup. The first control that is focsuable would get the focus automatically, when it was a picker it would open the selection automatically. 

In my case, I could solve this by adding a positive ZIndex value to the control I didn't want to have autofocus on. Now if I open and close the popup, the focus does not go back to the with the ZIndex.

No idea why this would work or what the side effects are, but I'm putting this out there in case it can help someone else.

**Comment by RichardMarsh** [comment]
> > Also experiencing this on Android 13 API Version 33.
> > This occurs if the Picker is the first visible control on a page and the page regains focus from another page or a popup disappearing.
> 
> I had the same problem when closing a popup. The first control that is focusable would get the focus automatically triggering the OnFocus event.
> 
> In my case, I could solve this by adding a positive ZIndex value to the control I didn't want to have autofocus on. Now if I open and close the popup, the focus does not go back to control with the ZIndex.
> 
> Focus goes to whichever control has a lower zindex.
> 
> No idea why this would work or what the side effects are, but I'm putting this out there in case it can help someone else.

While the above worked on a simple form I tried to apply this fix to a more complex form and for some reason it didn't work. However, my work around in this case was to set a zindex for all the focusable controls to the same posistive value and the autofocus would shift to the last entry on the page. 

Not ideal or a fix, but again, I thought this info might help someone else find a work around for their scenario

**Comment by RichardMarsh** [comment]
> > Also experiencing this on Android 13 API Version 33.
> > This occurs if the Picker is the first visible control on a page and the page regains focus from another page or a popup disappearing.
> 
> I had the same problem when closing a popup. The first control that is focusable would get the focus automatically triggering the OnFocus event.
> 
> In my case, I could solve this by adding a positive ZIndex value to the control I didn't want to have autofocus on. Now if I open and close the popup, the focus does not go back to control with the ZIndex.
> 
> Focus goes to whichever control has a lower zindex.
> 
> No idea why this would work or what the side effects are, but I'm putting this out there in case it can help someone else.


Found this work around does not work when controls are inside of a ScrollView

**Comment by RichardMarsh** [comment]
> @RichardMarsh, did you try my fix? It is working great for me.

Going to give it a bash soon

**Comment by smalgin** [comment]
@mchiz - many thanks, works like a charm!

**Comment by TKwakernaak** [comment]
@mchiz Thanks for your handler, it's working great

**Comment by akhilvswoodplc** [comment]
> @RichardMarsh, did you try my fix? It is working great for me.

Not working for me. Any changes required?

**Comment by corne-ac** [comment]
This is still a problem and workarounds are very inconsistent in working.

**Comment by uhbtl** [comment]
Same problem here. With the workaround the picker doesn´t automatically open anymore. But at the same time I cannot open the picker manually in code with the Focus method anymore.

**Comment by Andy-Donegan** [comment]
> Same problem here. With the workaround the picker doesn´t automatically open anymore. But at the same time I cannot open the picker manually in code with the Focus method anymore.

Read my suggestion before the work around, the original author Deserves Credit but I can not find the link to where I got it from. It is extremely basic but a fake picker invisible removes all of the problems and allows all of your other pickers to operate as normal. Not perfect but resolves your issues hopefully.

**Comment by thomasgalliker** [comment]
This seems to happen even on latest Android operating system. The steps I was able to reproduce it are:

1. Create a MainPage and DetailPage. Set MainPage of App to new NavigationPage(new MainPage).
2. Add an Entry to DetailPage.
3. On MainPage, place a Picker with some items... Make sure the Picker is the first focussable element in the visual tree.
4. Navigation to DetailPage.
5. Set focus to Entry by putting your finger into it.
6. Navigate back to MainPage.

**Actual outcome**
The Picker is automatically focussed (opened) when we navigate back from DetailPage to MainPage.

**Expected outcome**
The Picker is not opened when we navigate back to MainPage.

**Comment by uhbtl** [comment]
> Read my suggestion before the work around, the original author Deserves Credit but I can not find the link to where I got it from. It is extremely basic but a fake picker invisible removes all of the problems and allows all of your other pickers to operate as normal. Not perfect but resolves your issues hopefully.

You´re right, that works. Thank you.
Since I dont want to do that on every possible page, do you know a more generic way to implement this workaround?

**Comment by edisonjromor** [comment]
In NET MAUI 8.0.100 persist this bug

**Comment by raymondtrudeau** [comment]
Can confirm the workaround provided by @Andy-Donegan does work.  Couple of tweaks required - needed to increase the height of the fake entry - anything not visible did not seem to get the focus and as such was bypassed and went to the next control instead (picker).  Also, set the IsEnabled property to False to prevent clicking in the control.

Does anyone know though if this has been address in .net 9?

**Comment by holecekp** [comment]
This is a serious issue because the picker is opened automatically also on navigation back and tapping on it crashes the app with "System.InvalidOperationException: VirtualView cannot be null here".
